lte/gateway/python/magma/pipelined/app/middle.py

```diff
@@ -0,0 +1,189 @@
+        self._clean_restart = kwargs['config']['clean_restart']
+        self._msg_hub = MessageHub(self.logger)
+        self._datapath = None
+        self.tbl_num = self._midle_tbl_num
+
+    def _get_config(self, config_dict):
+        mtr_ip = None
+        mtr_port = None
+
+        if 'mtr_ip' in config_dict and 'mtr_interface' in config_dict and 'ovs_mtr_port_number' in config_dict:
+            self._mtr_service_enabled = True
+            mtr_ip = config_dict['mtr_ip']
+            mtr_port = config_dict['ovs_mtr_port_number']
+            mtr_mac = get_virtual_iface_mac(config_dict['mtr_interface'])
+        else:
+            mtr_ip = None
+            mtr_mac = None
+            mtr_port = None
+
+        li_port_name = None
+        if 'li_local_iface' in config_dict:
+            li_port_name = config_dict['li_local_iface']
+
+        return self.MiddleConfig(
+            mtr_ip=mtr_ip,
+            mtr_port=mtr_port,
+            li_port_name=li_port_name,
+            setup_type=config_dict.get('setup_type', None),
+            mtr_mac=mtr_mac,
+        )
+
+    def _get_default_flow_msgs(self, datapath) -> DefaultMsgsMap:
+        return {
+            self._midle_tbl_num: self._get_default_middle_flow_msgs(datapath),
+        }
+
+    def _get_default_middle_flow_msgs(self, dp):
+        """
+        Egress table is the last table that a packet touches in the pipeline.
+        Output downlink traffic to gtp port, uplink traffic to LOCAL
+
+        Raises:
+            MagmaOFError if any of the default flows fail to install.
+        """
+        msgs = []
+        next_tbl = self._service_manager.get_next_table_num(PHYSICAL_TO_LOGICAL)
+
+        # Allow passthrough pkts(skip enforcement and send to egress table)
+        ps_match = MagmaMatch(passthrough=PASSTHROUGH_REG_VAL)
+        msgs.append(
+            flows.get_add_resubmit_next_service_flow_msg(
+                dp,
+                self._midle_tbl_num, ps_match, actions=[],
+                priority=flows.PASSTHROUGH_PRIORITY,
+                resubmit_table=self._egress_tbl_num,
+            ),
+        )
+
+        match = MagmaMatch()
+        msgs.append(
+            flows.get_add_resubmit_next_service_flow_msg(
+                dp,
+                self._midle_tbl_num, match, actions=[],
+                priority=flows.DEFAULT_PRIORITY, resubmit_table=next_tbl,
+            ),
+        )
+
+        if self._mtr_service_enabled:
+            msgs.extend(
+                get_vlan_egress_flow_msgs(
+                    dp,
+                    self._midle_tbl_num,
+                    ether_types.ETH_TYPE_IP,
+                    self.config.mtr_ip,
+                    self.config.mtr_port,
+                    priority=flows.UE_FLOW_PRIORITY,
+                    direction=Direction.OUT,
+                    dst_mac=self.config.mtr_mac,
+                ),
+            )
+        return msgs
+
+    def _install_default_flows(self, datapath):
```

Comment:
function, _install_default_flows() is duplicated in ingress, middle and egress, can we move it to base calss.

Reply:
@pshelar: base.py is the super class of other controllers too. And those other controllers also have `_install_default_flows()` - but these are diverging implementations. Do we really want to implement `_install_default_flows()` of ingress, middle and egress in base.py and let all other controllers override this function?
Another approach would be to add an inheritance layer between base.py and ingress, middle and egress and put common logic there - this was already discussed with @koolzz above - this has the downside of adding inheritance complexity to the model.

Reply:
ok, thats fine.